[AMB-18177] webrtc_server crashes due to "Check failed: peer_connection_"

sambasevam opened the pull request:
Replace RTC_CHECK with a nullptr check instead to handle the peer_connection_ being closed when LocalDescription() is called.

Fix is untested, will run sanity check on host8.

davidtag commented on talk/owt/sdk/base/peerconnectionchannel.cc:
```diff
@@ -98,8 +98,11 @@ void PeerConnectionChannel::AddTransceiver(
 PeerConnectionChannel::LocalDescription() {
-  RTC_CHECK(peer_connection_);
-  return peer_connection_->local_description();
+  // Unfortunately this isn't entirely safe either as the scoped_refptr isn't
```
I don't think this code should change. This is essentially a read-only method on this class and the check makes sense here (like in neighboring methods). the problem is in the calling code. It's calling this method when the peer connection is closed, which points to a race condition. The solution should be in the calling code. See suggestion below.

sambasevam committed:
[AMB-18177] webrtc_server crashes due to "Check failed: peer_connection_"
Keep peer connection alive for LocalDescription() or return if it's not.
Also added a fix for AMB-16882 where we keep peer_connection_ alive for OnSignalingChange()

davidtag commented on talk/owt/sdk/p2p/p2ppeerconnectionchannel.cc:
```diff
@@ -761,6 +761,11 @@ void P2PPeerConnectionChannel::OnSetLocalSessionDescriptionSuccess() {
   // Setting maximum bandwidth here.
   ApplyBitrateSettings();
   auto desc = LocalDescription();
+  if (!desc) {
```
the problem is a race with `ClosePeerConnection()`. The problem is line 757-759 checks `ended_` under a lock, but once that lock is released, all bets are off. 

The solution we've used is the `GetPeerConnectionRef()` method. It internally checks the `ended_` flag and returns a reference counted version of the peer connection. That will guarantee the peer_connection_ doesn't become null before the method exits. You can simply replace line 757-760 with this method call and no other changes needed

sambasevam replied:
Oh, good point.

sambasevam replied:
I actually realized this as I was looking at [this bug](https://ambient-ai.atlassian.net/browse/AMB-16882), but you beat me to it. :-D

sambasevam replied:
Also, we still need to account for the peer connection ref being null, so I'll return if it's null.

sambasevam replied:
This seems to be a common problem though - I see the peer connection isn't checked elsewhere either. Is that intentional @davidtag? 

`  rtc::scoped_refptr<webrtc::PeerConnectionInterface> temp_pc_ = GetPeerConnectionRef();
`
`temp_pc_` could be null, but the caller assumes it's not.

davidtag replied:
not intentional, likely bugs

sambasevam committed:
Merge branch 'main' into AMB-18177-check-failed-peer_connection_
